File: SexyBiscuit.Engine/Debug/GraphicsBenchmark.cs
using SexyBiscuit.Engine.Core;
using SexyBiscuit.Engine.Rendering;

namespace SexyBiscuit.Engine.Debug;

/// <summary>
/// The result of a timed run: what the hardware actually managed.
/// </summary>
/// <remarks>
/// The low percentiles matter more than the average, and are the reason this reports
/// them. A game averaging 90 fps with a 1% low of 18 stutters visibly and is a worse
/// experience than a steady 60; an average alone cannot tell those apart, so a preset
/// chosen on the average alone would be chosen wrong.
/// </remarks>
public readonly record struct BenchmarkResult(
    float AverageFps,
    float OnePercentLowFps,
    float PointOnePercentLowFps,
    float FrameTimeStdDevMs,
    int   Frames,
    float Seconds,
    int   DrawCalls,
    int   Triangles,
    string SuggestedPreset)
{
    /// <summary>The one line the menu shows when a run finishes.</summary>
    public override string ToString()
        => $"{AverageFps:0} fps average · {OnePercentLowFps:0} fps 1% low · "
         + $"{PointOnePercentLowFps:0} fps 0.1% low · ±{FrameTimeStdDevMs:0.0} ms · "
         + $"{Triangles:N0} tris, {DrawCalls} draws · suggests {SuggestedPreset}";
}

/// <summary>
/// Measures what a machine can actually do, and maps the answer onto a preset.
/// </summary>
/// <remarks>
/// <para>
/// The mirror of <c>html5/src/debug/GraphicsBenchmark.js</c>. A pure accumulator: it is
/// handed a frame time each frame and is asked for a result when it has enough, so it
/// runs identically on a renderer it knows nothing about, and a test can drive it with
/// a list of numbers.
/// </para>
/// <para>
/// The first frames of a run are discarded. Starting a benchmark means switching preset,
/// which reallocates render targets, drops the primitive cache and recompiles nothing in
/// particular — and timing that would measure the switch rather than the hardware.
/// </para>
/// </remarks>
public sealed class GraphicsBenchmark
{
    /// <summary>Frames thrown away at the start, while the new preset settles.</summary>
    public const int WarmUpFrames = 30;

    private readonly List<float> _frameTimes = new();

    private float _elapsed;
    private int   _warmedUp;

    /// <summary>How long a run lasts, in seconds of real time after the warm-up.</summary>
    public float Duration { get; set; } = 8f;

    /// <summary>Whether a run is in progress.</summary>
    public bool IsRunning { get; private set; }

    /// <summary>0..1 through the current run, for a progress bar.</summary>
    public float Progress => IsRunning && Duration > 0f ? Math.Clamp(_elapsed / Duration, 0f, 1f) : 0f;

    /// <summary>The last completed run, or null.</summary>
    public BenchmarkResult? Result { get; private set; }

    /// <summary>The settings in force before the run, to be put back afterwards.</summary>
    public GraphicsSettings? Restore { get; set; }

    /// <summary>Starts a run, discarding anything a previous one collected.</summary>
    public void Start()
    {
        _frameTimes.Clear();
        _elapsed  = 0f;
        _warmedUp = 0;
        Result    = null;
        IsRunning = true;
    }

    /// <summary>Abandons a run without producing a result.</summary>
    public void Cancel()
    {
        IsRunning = false;
        _frameTimes.Clear();
    }

    /// <summary>
    /// Feeds one frame in. Returns the result on the frame the run finishes, else null.
    /// </summary>
    /// <param name="unscaledDeltaSeconds">
    /// Real seconds, never scaled ones: a game paused behind the menu has a time scale of
    /// zero, and a benchmark measured in scaled time would never advance.
    /// </param>
    /// <param name="stats">The renderer's counters, recorded with the result.</param>
    public BenchmarkResult? Tick(float unscaledDeltaSeconds, RenderStats? stats = null)
    {
        if (!IsRunning) return null;

        if (_warmedUp < WarmUpFrames) { _warmedUp++; return null; }

        // A frame long enough to be a stall — an alt-tab, a garbage collection the size
        // of a level load — is not this machine's frame rate and must not be averaged in.
        if (unscaledDeltaSeconds > 0f && unscaledDeltaSeconds < 1f)
        {
            _frameTimes.Add(unscaledDeltaSeconds);
            _elapsed += unscaledDeltaSeconds;
        }

        if (_elapsed < Duration) return null;

        IsRunning = false;
        Result    = Summarise(_frameTimes, stats);
        return Result;
    }

    // -------------------------------------------------------------------------
    // The arithmetic, which both engines do identically
    // -------------------------------------------------------------------------

    /// <summary>Turns a list of frame times into a result.</summary>
    public static BenchmarkResult Summarise(IReadOnlyList<float> frameTimes, RenderStats? stats = null)
    {
        if (frameTimes.Count == 0)
            return new BenchmarkResult(0f, 0f, 0f, 0f, 0, 0f, 0, 0,
                                       GraphicsSettings.Presets.Suggest(0f, 0f));

        var sorted = new List<float>(frameTimes);
        sorted.Sort();          // Ascending, so the slowest frames are at the end.

        float total = 0f;
        foreach (float t in frameTimes) total += t;

        float average = total / frameTimes.Count;
        float averageFps = average > 0f ? 1f / average : 0f;

        float variance = 0f;
        foreach (float t in frameTimes) variance += (t - average) * (t - average);
        variance /= frameTimes.Count;

        float onePercentLow = LowFps(sorted, 0.01f);

        return new BenchmarkResult(
            AverageFps:            averageFps,
            OnePercentLowFps:      onePercentLow,
            PointOnePercentLowFps: LowFps(sorted, 0.001f),
            FrameTimeStdDevMs:     MathF.Sqrt(variance) * 1000f,
            Frames:                frameTimes.Count,
            Seconds:               total,
            DrawCalls:             stats?.DrawCalls ?? 0,
            Triangles:             stats?.Triangles ?? 0,
            SuggestedPreset:       GraphicsSettings.Presets.Suggest(averageFps, onePercentLow));
    }

    /// <summary>
    /// The average frame rate over the slowest fraction of frames.
    /// </summary>
    /// <remarks>
    /// The average of the tail, not the single worst frame at that percentile. One frame
    /// is noise — a scheduler hiccup lands there as readily as a real stall — and the
    /// average over the tail is what a player experiences as the game hitching.
    /// </remarks>
    private static float LowFps(List<float> ascending, float fraction)
    {
        int count = Math.Max(1, (int)MathF.Ceiling(ascending.Count * fraction));

        float total = 0f;
        for (int i = ascending.Count - count; i < ascending.Count; i++) total += ascending[i];

        float mean = total / count;
        return mean > 0f ? 1f / mean : 0f;
    }
}

File: html5/src/debug/GraphicsBenchmark.js
// -----------------------------------------------------------------------------
// GraphicsBenchmark — measures what a machine can actually do, and maps the answer
// onto a preset.
//
// The mirror of SexyBiscuit.Engine/Debug/GraphicsBenchmark.cs.
//
// A pure accumulator: it is handed a frame time each frame and asked for a result
// when it has enough, so it runs identically on a renderer it knows nothing about and
// a test can drive it with a list of numbers.
//
// The low percentiles matter more than the average, which is why they are reported. A
// game averaging 90 fps with a 1% low of 18 stutters visibly and is a worse experience
// than a steady 60; an average alone cannot tell those apart, so a preset chosen on
// the average alone would be chosen wrong.
// -----------------------------------------------------------------------------

import { GraphicsSettings } from '../rendering/GraphicsSettings.js';

/** Frames thrown away at the start, while the new preset settles. */
export const WARM_UP_FRAMES = 30;

export class GraphicsBenchmark {
    constructor() {
        /** How long a run lasts, in seconds of real time after the warm-up. */
        this.duration = 8;

        /** Whether a run is in progress. */
        this.isRunning = false;

        /** The last completed run, or null. */
        this.result = null;

        /** The settings in force before the run, to be put back afterwards. */
        this.restore = null;

        this._frameTimes = [];
        this._elapsed = 0;
        this._warmedUp = 0;
    }

    /** 0..1 through the current run, for a progress bar. */
    get progress() {
        if (!this.isRunning || this.duration <= 0) return 0;
        return Math.min(1, Math.max(0, this._elapsed / this.duration));
    }

    /** Starts a run, discarding anything a previous one collected. */
    start() {
        this._frameTimes = [];
        this._elapsed = 0;
        this._warmedUp = 0;
        this.result = null;
        this.isRunning = true;
    }

    /** Abandons a run without producing a result. */
    cancel() {
        this.isRunning = false;
        this._frameTimes = [];
    }

    /**
     * Feeds one frame in. Returns the result on the frame the run finishes, else null.
     *
     * Real seconds, never scaled ones: a game paused behind the menu has a time scale
     * of zero, and a benchmark measured in scaled time would never advance.
     */
    tick(unscaledDeltaSeconds, stats = null) {
        if (!this.isRunning) return null;

        // The first frames of a run are discarded. Starting a benchmark means switching
        // preset, which reallocates buffers and drops the primitive cache, and timing
        // that would measure the switch rather than the hardware.
        if (this._warmedUp < WARM_UP_FRAMES) { this._warmedUp++; return null; }

        // A frame long enough to be a stall -- a hidden tab, a garbage collection the
        // size of a level load -- is not this machine's frame rate.
        if (unscaledDeltaSeconds > 0 && unscaledDeltaSeconds < 1) {
            this._frameTimes.push(unscaledDeltaSeconds);
            this._elapsed += unscaledDeltaSeconds;
        }

        if (this._elapsed < this.duration) return null;

        this.isRunning = false;
        this.result = summarise(this._frameTimes, stats);
        return this.result;
    }
}

// -----------------------------------------------------------------------------
// The arithmetic, which both engines do identically
// -----------------------------------------------------------------------------

/** Turns a list of frame times into a result. */
export function summarise(frameTimes, stats = null) {
    if (frameTimes.length === 0) {
        return {
            averageFps: 0, onePercentLowFps: 0, pointOnePercentLowFps: 0,
            frameTimeStdDevMs: 0, frames: 0, seconds: 0,
            drawCalls: 0, triangles: 0,
            suggestedPreset: GraphicsSettings.suggestFor(0, 0),
        };
    }

    const sorted = [...frameTimes].sort((a, b) => a - b);   // Slowest frames last.
    const total = frameTimes.reduce((sum, t) => sum + t, 0);
    const average = total / frameTimes.length;
    const averageFps = average > 0 ? 1 / average : 0;

    const variance = frameTimes.reduce((sum, t) => sum + (t - average) ** 2, 0) / frameTimes.length;

    const onePercentLowFps = lowFps(sorted, 0.01);

    return {
        averageFps,
        onePercentLowFps,
        pointOnePercentLowFps: lowFps(sorted, 0.001),
        frameTimeStdDevMs: Math.sqrt(variance) * 1000,
        frames: frameTimes.length,
        seconds: total,
        drawCalls: stats?.drawCalls ?? 0,
        triangles: stats?.triangles ?? 0,
        suggestedPreset: GraphicsSettings.suggestFor(averageFps, onePercentLowFps),
    };
}

/** The one line the menu shows when a run finishes. */
export function describe(result) {
    return `${Math.round(result.averageFps)} fps average · ${Math.round(result.onePercentLowFps)} fps 1% low · `
         + `${Math.round(result.pointOnePercentLowFps)} fps 0.1% low · ±${result.frameTimeStdDevMs.toFixed(1)} ms · `
         + `${result.triangles.toLocaleString('en-GB')} tris, ${result.drawCalls} draws · `
         + `suggests ${result.suggestedPreset}`;
}

/**
 * The average frame rate over the slowest fraction of frames.
 *
 * The average of the tail, not the single worst frame at that percentile. One frame is
 * noise -- a scheduler hiccup lands there as readily as a real stall -- and the average
 * over the tail is what a player experiences as the game hitching.
 */
function lowFps(ascending, fraction) {
    const count = Math.max(1, Math.ceil(ascending.length * fraction));
    const tail = ascending.slice(ascending.length - count);
    const mean = tail.reduce((sum, t) => sum + t, 0) / count;
    return mean > 0 ? 1 / mean : 0;
}
